lib/iex/lib/iex/evaluator.ex

```diff
@@ -225,21 +219,21 @@ defmodule IEx.Evaluator do
     end
   end
 
-  defp do_eval(@break_trigger, %IEx.State{cache: ''} = iex_state, state) do
-    {iex_state, state}
-  end
-
-  defp do_eval(@break_trigger, iex_state, _state) do
-    :elixir_errors.parse_error([line: iex_state.counter], "iex", "incomplete expression", "")
-  end
-
   defp do_eval(latest_input, iex_state, state) do
     code = iex_state.cache ++ latest_input
     line = iex_state.counter
     put_history(state)
     put_whereami(state)
-    quoted = Code.string_to_quoted(code, line: line, file: "iex")
-    handle_eval(quoted, code, line, iex_state, state)
+    {parser_module, parse_method, additional_args} = IEx.Config.parser()
+    arguments = [latest_input, [line: line, file: "iex"], iex_state.cache | additional_args]
+
+    case apply(parser_module, parse_method, arguments) do
+      {:ok, forms, ""} ->
+        handle_eval({:ok, forms}, code, line, iex_state, state)
+
+      {:incomplete, _input} ->
+        handle_eval(:incomplete, code, line, iex_state, state)
+    end
```

Comment:
This should rather be something like:

```suggestion
    case apply(parser_module, parse_method, arguments) do
      {:ok, forms, ""} ->
        iex_state = %{iex_state | buffer: buffer, counter: iex_state.counter + 1}
        {iex_state, handle_eval(forms, code, line, state)}

      {:incomplete, buffer} ->
        {%{iex_state | buffer: buffer}, state}
    end
```

ANd then `handle_eval` can be simplified a bit. :)

Reply:
hi @josevalim  @jrusso1020 , some parsers may be oriented more to give side effects than produce any sensible elixir values (they even may not have much common with elixir). Returning `IEx.dont_display_result/0` does it, but bringing it to API would not be nice. Maybe consider to replace `:ok` tuple with `:value` and `:no_value` result tuples (you may propose better names)?
Seems the parser feature will become available sooner than later. :+1:

Reply:
I would prefer to explicitly return `IEx.dont_display_result` for now. Especially because I think you should avoid actually doing work in the parser. Instead you should return quoted expressions.